src/twisted/internet/asyncfs.py

```diff
@@ -0,0 +1,709 @@
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+# -*- test-case-name: twisted.internet.test.test_async -*-
+"""
+This module contains asynchronous filesystem interfaces and reference implementations
+"""
+
+from zope.interface import Interface, Attribute, implementer
+from twisted.internet import reactor
+from twisted.internet.interfaces import (
+    IPushProducer,
+    IProducer,
+    IConsumer,
+)
+from twisted.python.threadpool import ThreadPool
+from twisted.internet.threads import (
+    deferToThreadPool,
+    blockingCallFromThread,
+)
+from twisted.internet.defer import Deferred
+from twisted.logger import Logger
+from typing import Tuple, Iterable, Optional, Callable, cast, Any
+import os
+import os.path
+import threading
+
+log = Logger()
+
+
+DEFAULT_BUFFER_MAX = 2 ** 12  # four megabytes
+DEFAULT_CHUNK_SIZE = 4096
+
+
+class IAsyncFilesystem(Interface):
+
+    """
+    an asynchronous filesystem
+    based closely on L{twisted.conch.interfaces.ISFTPServer}
+    """
+
+    def openFile(
+        filename: str, flags: int = 0, attrs: Optional[dict] = None
+    ) -> "Deferred":
+        """
+        Called when the clients asks to open a file.
+
+        @param filename: a string representing the file to open.
+
+        @param flags: an integer of the flags to open the file with, ORed
+        together.  The flags and their values are passed directly to L{os.open}
+
+        @param attrs: a list of attributes to open the file with.  It is a
+        dictionary, consisting of 0 or more keys.  The possible keys are::
+
+            size: the size of the file in bytes
+            uid: the user ID of the file as an integer
+            gid: the group ID of the file as an integer
+            permissions: the permissions of the file with as an integer.
+            the bit representation of this field is defined by POSIX.
+            atime: the access time of the file as seconds since the epoch.
+            mtime: the modification time of the file as seconds since the epoch.
+            ext_*: extended attributes.  The server is not required to
+            understand this, but it may.
+
+        NOTE: there is no way to indicate text or binary files.
```

Comment:
So all files are text?  All files are binary?  Files type is automatically detected?  File type is platform dependent?  Filesystem dependent?